File: GestureCreator/Classes/Utilities/DeferredActionManager.m
//
// Created by Timur Turaev on 18.08.14.
// Copyright (c) 2014 Yandex. All rights reserved.
//

#import "DeferredActionManager.h"

@interface DeferredActionManager ()
@property(nonatomic, readonly) NSMutableDictionary *deferredBlocks;
@end

@implementation DeferredActionManager

- (instancetype)init {
    self = [super init];
    if (self) {
        _deferredBlocks = [NSMutableDictionary dictionary];
    }

    return self;
}

- (void)performActionAfterDelay:(NSTimeInterval)delay withKey:(NSString *)key block:(DeferredBlock)block {
    [self removeActionForKey:key];
    self.deferredBlocks[key] = block;
    [self performSelector:@selector(fireBlockWithKey:) withObject:key afterDelay:delay];
}

- (void)fireBlockWithKey:(NSString *)key {
    DeferredBlock block = self.deferredBlocks[key];
    if (block) {
        block();
    }
    [self removeActionForKey:key];
}

- (void)removeActionForKey:(NSString *)key {
    [self.deferredBlocks removeObjectForKey:key];
    [self.class cancelPreviousPerformRequestsWithTarget:self selector:@selector(fireBlockWithKey:) object:key];
}

- (void)removeAllActions {
    [self.deferredBlocks removeAllObjects];
    [self.class cancelPreviousPerformRequestsWithTarget:self];
}

@end

File: GestureCreator/Classes/Utilities/DeferredActionManager.swift
//
//  DeferredActionManager.swift
//  GestureCreator
//
//  Created by Тимур on 08.04.15.
//  Copyright (c) 2015 Yandex LLC. All rights reserved.
//

import Foundation

//class DeferredActionManager : NSObject {
//    typealias DeferredAction = () -> ()
//    private var deferredActions = [String:DeferredAction]()
//    
//    func perform(afterDelay delay: NSTimeInterval, withKey key: String, action: DeferredAction) {
//        self.removeActionForKey(key)
//        deferredActions[key] = action
//        self perform action with delay ?????????????????
//    }
//    
//    func removeActionForKey(key: String) {
//        deferredActions.removeValueForKey(key)
//        DeferredActionManager.cancelPreviousPerformRequestsWithTarget(self, selector: Selector("runActionWithKey:"), object: key)
//    }
//    
//    func removeAllDeferredActions() {
//        deferredActions.removeAll(keepCapacity: false)
//        DeferredActionManager.cancelPreviousPerformRequestsWithTarget(self)
//    }
//    
//    private func runActionWithKey(key: String) {
//        if let action = deferredActions[key] {
//            action()
//            deferredActions.removeValueForKey(key)
//        }
//    }
//}